Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,158 @@
+﻿namespace Microsoft.ApplicationInsights.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Globalization;
+    using System.IO;
+    using System.Net;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A store for instrumentation App Ids. This makes sure we don't query the public endpoint to find an app Id for the same Ikey more than once.
+    /// </summary>
+    internal static class CorrelationIdLookupHelper
+    {
+        /// <summary>
+        /// Max number of app ids to cache.
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        private const int GET_APP_ID_TIMEOUT = 2000; // milliseconds
+
+        private const string CORRELATION_ID_FORMAT = "aid-v1:{0}";
+
+        private const string APPID_QUERY_API_RELATIVE_URI_FORMAT = "api/profiles/{0}/appId";
+
+        private static ConcurrentDictionary<string, string> knownCorrelationIds = new ConcurrentDictionary<string, string>();
+
+        private static Func<string, string, Task<string>> provideAppId = FetchAppIdFromService;
+
+        /// <summary>
+        /// This is a test hook. Use this to provide your own test implementation for fetching the appId.
+        /// </summary>
+        /// <param name="appIdProviderMethod">The delegate to be called to fetch the appId</param>
+        public static void OverrideAppIdProvider(Func<string, string, Task<string>> appIdProviderMethod)
+        {
+            provideAppId = appIdProviderMethod;
+        }
+
+        /// <summary>
+        /// Retrieves the correlation id corresponding to a given instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key string.</param>
+        /// <param name="breezeEndpointAddress">Breeze endpoint to query against.</param>
+        /// <param name="correlationId">AppId corresponding to the provided instrumentation key</param>
+        /// <returns>true if correlationId was successfully retrieved; false otherwise.</returns>
+        public static bool TryGetXComponentCorrelationId(string instrumentationKey, string breezeEndpointAddress, out string correlationId)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                throw new ArgumentNullException("instrumentationKey");
+            }
+
+            var found = knownCorrelationIds.TryGetValue(instrumentationKey, out correlationId);
+
+            if (found)
+            {
+                return true;
+            }
+            else
+            {
+
+                if (breezeEndpointAddress == null)
+                {
+                    throw new ArgumentNullException("breezeEndpointAddress");
```

Comment:
Actually this shouldn't throw anyways because the method signature is Try... Updated.